### PR TITLE
spring config

### PR DESCRIPTION
**Pull request by oarozu**
Spring config and Book CRUD

image

### CODE REVIEW

File: src/main/java/wolox/training/controllers/BookController.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    @GetMapping
+    public Iterable findAll() {
+        return bookRepository.findAll();
+    }
+
+    @GetMapping("/title/{bookTitle}")
+    public List findByTitle(@PathVariable String bookTitle) {
+        try {
+            return bookRepository.findByTitle(bookTitle);
```
Comment (cramallo) [inline src/main/java/wolox/training/controllers/BookController.java]: You can move this code to a service class, in this case "BookService" class. It's a good practice to call services from the controller, and call the repositories from the services. Check this post https://medium.com/karibu-blog/c%C3%B3mo-aplicar-jpa-en-spring-boot-e45ac5bee3b4

File: .gradle/buildOutputCleanup/cache.properties [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#Mon Jun 29 12:01:20 ART 2020
```
Comment (facucompe) [inline .gradle/buildOutputCleanup/cache.properties]: All `.gradle` should not be in the remote repository.

File: .idea/.name [hunk 3]
```diff
@@ -0,0 +1 @@
+training
```
Comment (facucompe) [inline .idea/.name]: Same here. All `.idea` files should not be in the remote repository

File: src/main/java/wolox/training/repositories/BookRepository.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import org.springframework.stereotype.Repository;
+import wolox.training.models.Book;
+
+import java.util.List;
+
+
+@Repository
+public interface BookRepository extends CrudRepository<Book, String> {
+
+    public Book findById(Long id);
```
Comment (facucompe) [inline src/main/java/wolox/training/repositories/BookRepository.java]: No need to use `public`.

Comment (facucompe) [reply src/main/java/wolox/training/repositories/BookRepository.java]: Also, try returning `Optional<Book>`

**Review by facucompe — CHANGES_REQUESTED** [review]

File: src/main/java/wolox/training/controllers/BookController.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        } catch (BookNotFoundException ex) {
+            throw new ResponseStatusException(
+                    HttpStatus.NOT_FOUND, "Book Not Found", ex);
+        }
+    }
+
+    @GetMapping("/{id}")
+    public Book findOne(@PathVariable Long id) {
```
Comment (cramallo) [inline src/main/java/wolox/training/controllers/BookController.java]: Use final here, and in all variables that won't change. Check the benefits of using final https://softwareengineering.stackexchange.com/questions/98691/excessive-use-final-keyword-in-java

File: src/main/java/wolox/training/controllers/BookController.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package wolox.training.controllers;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.*;
```
Comment (cramallo) [inline src/main/java/wolox/training/controllers/BookController.java]: Try to avoid using *.

**Review by cramallo — CHANGES_REQUESTED** [review]

File: src/main/java/wolox/training/controllers/BookController.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public Book create(@RequestBody Book book) {
+        return bookRepository.save(book);
+    }
+
+    @DeleteMapping("/{id}")
+    public void delete(@PathVariable Long id) {
+        bookRepository.findById(id);
```
Comment (cramallo) [inline src/main/java/wolox/training/controllers/BookController.java]: check this, this is not necessary

Comment (cramallo) [reply src/main/java/wolox/training/controllers/BookController.java]: also you can add a try and catch, to catch exceptions

**Review by cramallo — CHANGES_REQUESTED** [review]

**Review by cramallo — COMMENTED** [review]
Try to make a pr by card

File: src/main/java/wolox/training/controllers/BookController.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    @Autowired
+    private BookRepository bookRepository;
+
+    @GetMapping
+    public Iterable findAll() {
+        return bookRepository.findAll();
+    }
+
+    @GetMapping("/title/{bookTitle}")
+    public List findByTitle(@PathVariable String bookTitle) {
```
Comment (cramallo) [inline src/main/java/wolox/training/controllers/BookController.java]: You should indicate what you are returning, List of something(using <>)

**Review by cramallo — CHANGES_REQUESTED** [review]

**Review by cramallo — COMMENTED** [review]
Keep in mind that it is common for controllers to return a DTO, not directly the entity. https://www.oscarblancarteblog.com/2018/11/30/data-transfer-object-dto-patron-diseno/

File: src/main/java/wolox/training/controllers/BookController.java (hunk 6 above)
Comment (facucompe) [reply src/main/java/wolox/training/controllers/BookController.java]: 👍 Follow this to configure intellij https://intellij-support.jetbrains.com/hc/en-us/community/posts/206203659-Turn-off-Wildcard-imports-

File: src/main/java/wolox/training/controllers/BookController.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import wolox.training.services.BookService;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/books")
+public class BookController {
+
+    @Autowired
+    private BookRepository bookRepository;
```
Comment (cramallo) [inline src/main/java/wolox/training/controllers/BookController.java]: you should remove this from the controller and inject only in the service

File: src/main/java/wolox/training/controllers/BookController.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+    @GetMapping("/{id}")
+    public final Book findOne(@PathVariable Long id) {
+        return bookRepository.findById(id).orElseThrow(BookNotFoundException::new);
+    }
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public Book create(@RequestBody Book book) {
+        return bookRepository.save(book);
```
Comment (cramallo) [inline src/main/java/wolox/training/controllers/BookController.java]: move this line to the service class

File: src/main/java/wolox/training/controllers/BookController.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @DeleteMapping("/{id}")
+    public void delete(@PathVariable Long id) {
+        bookRepository.findById(id)
```
Comment (cramallo) [inline src/main/java/wolox/training/controllers/BookController.java]: same

File: src/main/java/wolox/training/controllers/BookController.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+
+    }
+
+    @PutMapping("/{id}")
+    public Book updateBook(@RequestBody Book book, @PathVariable Long id) {
+        if (book.getId() != id) {
+            throw new BookIdMismatchException();
+        }
+        bookRepository.findById(id).orElseThrow(BookNotFoundException::new);
```
Comment (cramallo) [inline src/main/java/wolox/training/controllers/BookController.java]: same

**Review by cramallo — CHANGES_REQUESTED** [review]

**Review by cramallo — APPROVED** [review]

**Pull request merged**